Test types are importable in CI

schneems opened the pull request:
This PR adds a typescript file and then it exercises it to ensure that types are exported correctly via top level `npm run build`. In theory if something major is wrong with either the build process or the `src/index.ts` then this smoke test has a chance of catching it.

Closes GUS-W-9649147.

schneems committed:
Test types are importable

This PR adds a typescript file and then it exercises it to ensure that types are exported correctly via top level `npm run build`. In theory if something major is wrong with either the build process or the `src/index.ts` then this smoke test has a chance of catching it.

schneems committed:
Merge branch 'main' into schneems/test-me

edmorley commented on .gitignore:
```diff
@@ -14,3 +14,5 @@ npm-debug.log
 *.log
 
 /.nyc_output/
+node_modules/
```
There's a `/node_modules` entry higher up this file, could you edit that one / combine with this line?

schneems replied:
I removed mine and edited the one above. Sharp eye. Thanks

edmorley commented on test/fixtures/typescript-function/package.json:
```diff
@@ -0,0 +1,15 @@
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
```suggestion
    "build": "tsc"
```

edmorley commented on .circleci/config.yml:
```diff
@@ -26,3 +26,7 @@ jobs:
+
+      - run:
+          name: Test types are built
+          command: cd test/fixtures/typescript-function && npm install && npx tsc
```
This uses the new `build` scripts target added in the review comment below. 

```suggestion
          command: cd test/fixtures/typescript-function && npm install && npm run build
```

edmorley commented on test/fixtures/typescript-function/src/index.ts:
```diff
@@ -0,0 +1,5 @@
+import { ReferenceId } from "sf-fx-sdk-nodejs"
+
+console.log("Here is code!");
```
Could you make this file be an example hello world function, representative of what a customer might be using?

schneems replied:
Sounds good. I frankly don't know what one looks like. I was trying to go with the maximum separation of concerns. I'm waiting on Lillian to publish the template and I'll modify that so it can run with little/no context and no salesforce access needed.

schneems replied:
Updated. I also really like your suggestion of having circle run the tests in the project. I moved the fixture to a top level "fixtures" directory. This helps avoid future mocha collisions from having nested test files under "test". I also had it run linting and found two issues with the source template code.

schneems committed:
Update test/fixtures/typescript-function/package.json

Co-authored-by: Ed Morley <501702+edmorley@users.noreply.github.com>

schneems committed:
Update .circleci/config.yml

Co-authored-by: Ed Morley <501702+edmorley@users.noreply.github.com>

schneems committed:
Resolve multiple node_modules in .gitignore

Node modules is listed several times in .gitignore. This declaration handles all our current cases.

schneems committed:
Test using typescript template

The prior test just checked that types import. This test includes a proper typescript salesforce function template. In addition it also executes the tests and linter against the project.

edmorley commented on .circleci/config.yml:
```diff
@@ -26,3 +26,7 @@ jobs:
+
+      - run:
+          name: Test types are built
+          command: cd fixtures/sample-typescript-project && npm install && npm run build lint test
```
I'm pretty sure `npm run` can only run one command at a time

edmorley replied:
^ outdated with latest force push

edmorley commented on .gitignore:
```diff
@@ -14,3 +14,4 @@ npm-debug.log
 *.log
 
 /.nyc_output/
+test/fixtures/*/dist/**
```
This needs updating for the new fixture location, and the committed `dist` files from the last commit to be deleted.

This could also be combined with the repo-root `/dist` entry on line 10 in this file? :-)

schneems commented:
All comments resolved. All tests passing. This is a much stronger PR thanks to your help! Thank you!

edmorley reviewed (commented):
This is great thank you!

I think these files still need deleting?
`fixtures/sample-typescript-project/dist/index.js`
`fixtures/sample-typescript-project/dist/index.js.map`

The `/dist` entry on line 10 in `.gitignore` also needs changing to `dist/` so it excludes both the repo root dist, and the fixtures build dist.

edmorley committed:
Remove committed dist/ and adjust .gitignore to exclude it

edmorley reviewed (approved):
I pushed a commit to remove the dist directory and updated .gitignore, so we can merge this and close out the TypeScript epic :-)

schneems commented:
Great, thanks!
